### ISSUE TITLE
Testing Schematron - regular expressions inside fn:matches() generate an error in the preprocessed Schematron

### ISSUE THREAD

**Issue by AlexJitianu** [xspec/xspec#1618]
This is a situation reported by one of our common users. We have this Schematron file. Please note the last commented assert. Once you un-comment this the issue reproduces.

```
<?xml version="1.0" encoding="UTF-8"?>
<sch:schema
    queryBinding="xslt2"
    xmlns:sch=["http://purl.oclc.org/dsdl/schematron"](http://purl.oclc.org/dsdl/schematron)
    xmlns:sqf=["http://www.schematron-quickfix.com/validator/process"](http://www.schematron-quickfix.com/validator/process)>
    <sch:pattern>
        <sch:rule
            context="*"
            id="regex">

            <sch:assert
                id="test1"
                test="true()">phrase</sch:assert>

            <sch:let
                name="expr"
                value="'(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))'" />
            <sch:assert
                id="test2"
                test="matches('::', $expr)">phrase</sch:assert>

            <!--<sch:assert
                id="test3"
                test="matches('::', '(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))')">phrase</sch:assert>-->

        </sch:rule>
    </sch:pattern>
</sch:schema> 
```

The XSpec file:

```
<?xml version="1.0" encoding="UTF-8"?>
<x:description
    schematron="regex.sch"
    xmlns:x=["http://www.jenitennison.com/xslt/xspec"](http://www.jenitennison.com/xslt/xspec)>
    <x:scenario
        label="regex">
        <x:context>
            <element />
        </x:context>
        <x:expect-not-assert
            id="test1"
            label="test1" />
        <x:expect-not-assert
            id="test2"
            label="test2" />
    </x:scenario>
</x:description>
```

The last assert in the Shematron contains a {1,} which generates an error _(Unexpected token "}" at start of expression)_ in the preprocessed sch in the value of `svrl:failed-assert/@test` because it is mistaken for an attribute value template.

```
<svrl:failed-assert xmlns:svrl=["http://purl.oclc.org/dsdl/svrl"](http://purl.oclc.org/dsdl/svrl)
                    test="matches('::', '(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))')">
                    <xsl:attribute name="id">test3</xsl:attribute>
                    <xsl:attribute name="location">
                        <xsl:apply-templates select="." mode="schematron-select-full-path"/>
                    </xsl:attribute>
                    <svrl:text>phrase</svrl:text>
                </svrl:failed-assert>
```

I'm not entirely sure, but perhaps these values should be escaped before being copied inside `svrl:failed-assert/@test`.

**Comment by galtm** [comment]
@AlexJitianu, thanks for passing this along and for the hints about the cause of the problem.

**Comment by galtm** [comment]
Fixed by #1626

**Comment by AirQuick** [comment]
Looks like this is Schematron/schematron#85.
Not repro with the `schxslt` branch. (Currently SchXslt 1.8.6).

**Comment by AirQuick** [comment]
The _skeleton_ implementation has a lot of bugs. In general, I don't think XSpec should take responsibility for such many bugs in `lib/iso-schematron/`.

**Comment by galtm** [comment]
> Fixed by #1626

**Update:** Although #1626 the third-party "skeleton implementation" of Schematron between the released XSpec v2.2 and the future XSpec v2.3 to work around this issue, the patch has been reverted in #1713. As a result, this problem will not go away in XSpec v2.3.

A related issue, https://github.com/18F/fedramp-automation/issues/396, mentioned a workaround in the user's schema code. For other situations where a user experiences this bug in the skeleton implementation, other options the user can consider are to fix the bug privately in their own copy of the skeleton implementation of Schematron, or switch to the SchXslt implementation of Schematron (https://github.com/schxslt/schxslt).
